File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.cpp
// User function template for C++

class Solution {
  public:
    int longestKSubstr(string &s, int k) {
        // your code here
        
        int n = s.size();
        int i = 0;
        int j = 0;
        unordered_map<char,int> ump;
        int ans = -1;
        
        while(j<n) {
            ump[s[j]]++;
            
            if(ump.size()>k) {
                ump[s[i]]--;
                if(ump[s[i]]==0) ump.erase(s[i]);
                i++;
            }
            if(ump.size()==k) {
                ans = max(ans,j-i+1);
            }
            j++;
        }
        
        return ans;
        
    }
};

File: Difficulty: Medium/Longest Substring with K Uniques/longest-substring-with-k-uniques.java
class Solution {
    public int longestKSubstr(String s, int k) {
        // code here
        int n = s.length();
        int[] freq = new int[26];
        
       // for(int i = 0;i<n;i++) freq[i] = -1;
        
        int i = 0;
        int j = 0;
        int cnt = 0;
        int ans = -1;
        
        while(j<n) {
            
            if(freq[s.charAt(j)-'a']==0) {
                cnt++;
            }
            freq[s.charAt(j)-'a']++;
            if(cnt>k) {
                freq[s.charAt(i)-'a']--;
                if(freq[s.charAt(i)-'a']==0) {
                    cnt--;
                }
                i++;
                
            }
            
            
            
            if(cnt==k) {
                ans = Math.max(ans,j-i+1);
            }
            j++;
        }
        
        return ans;
    }
}